refactor(api): Add move splitting capability to driver move

sfoster1 opened the pull request:
This is in service of unsticking frozen plunger axes.

When multi pipettes of certain models sit idle for too long, their plunger o
rings create more friction the next time they try to move. To move the o rings,
we have to move a short distance slowly with a higher current before continuing
the rest of the motion.

While this could happen at a higher level, implementing this as splitting up a
move commanded for other reasons lets us more easily share the logic between the
v1 and v2 hardware handling code. Now, the hardware handling code has to check
how long since the plunger last moved and command the split if necessary (or, if
the parameters for the split end up meaning that not that much time elapses
during the slow move, just do it every time in some contexts).

Closes #4674

## Testing
Check out the implementation and the tests and perhaps run it on a robot; at this point however the code isn't used anywhere

sfoster1 committed:
refactor(api): Add move splitting capability to driver move

This is in service of unsticking frozen plunger axes.

When multi pipettes of certain models sit idle for too long, their plunger o
rings create more friction the next time they try to move. To move the o rings,
we have to move a short distance slowly with a higher current before continuing
the rest of the motion.

While this could happen at a higher level, implementing this as splitting up a
move commanded for other reasons lets us more easily share the logic between the
v1 and v2 hardware handling code. Now, the hardware handling code has to check
how long since the plunger last moved and command the split if necessary (or, if
the parameters for the split end up meaning that not that much time elapses
during the slow move, just do it every time in some contexts).

Closes #4674

sfoster1 commented:
Adding do not merge because i realized i forgot to do the speed part of this!

sfoster1 committed:
fixup: add speed setting

sfoster1 commented:
OK added the speed control stuff.

Laura-Danielle commented on api/src/opentrons/drivers/smoothie_drivers/driver_3_0.py:
```diff
@@ -1268,87 +1282,158 @@ def _write_to_pipette(self, gcode: str, mount: str, data_string: str):
                 (axis in DISABLE_AXES) or
-                (coords is None) or
-                isclose(coords, self.position[axis],
+                isclose(coord, self.position[axis],
```
The resolution of our XY axis is approximately .4mm, should we maybe bump up the tolerances here? Right now this would only allow for about a ~0.034 margin.

sfoster1 replied:
That is probably a good idea! As far as I know these parameters were chosen arbitrarily.

Laura-Danielle commented on api/src/opentrons/drivers/smoothie_drivers/driver_3_0.py:
```diff
@@ -1268,87 +1282,158 @@ def _write_to_pipette(self, gcode: str, mount: str, data_string: str):
+
+        The function may issue up to 3 moves:
+        - if move splitting is required, the split move
+        - the actual move, plus a bit extra to give room to preload backlash
```
Should we not just subtract out the known backlash instead of preloading?

sfoster1 replied:
We really should be doing both - to preload, move a little farther and then move back but don't change the position. This method, however, is just the way we've always been doing it, just with some more comments

sfoster1 committed:
fixup: push timing down to the driver

Unfortunately, move timing is reliant on checking the last motion _after_ any
possible return from pause. Isn't that cool? So it has to live in the driver.

Laura-Danielle reviewed (approved):
Looks good to me 👍
